`concept proxiable` potentially depend on itself when using `facade_aware_overload_t`

The following code should compile, but it won't.

```cpp
#include "proxy.h"

template <class It, class F>
bool operator==(const It& it, const pro::proxy<F>& rhs) noexcept
    requires(!std::is_same_v<It, pro::proxy<F>>) {
  return typeid(It) == proxy_typeid(rhs) && it == proxy_cast<const It&>(rhs);
}
template <class F>
using SelfComparisonOverload = bool(const pro::proxy<F>& rhs) const noexcept;

template <class T>
struct Iterator : pro::facade_builder
    ::support_direct_rtti
    ::add_direct_convention<pro::operator_dispatch<"++">, void() noexcept>
    ::add_direct_convention<pro::operator_dispatch<"!=">, pro::facade_aware_overload_t<SelfComparisonOverload>>
    ::add_convention<pro::implicit_conversion_dispatch, T&() const noexcept>
    ::build {};

int main() {
  int a[3];
  pro::proxy<Iterator<int>> p = a;  // <-- Won't compile
}
```

Error message from Clang:

```cpp
In file included from 1.cc:1:
proxy.h:757:5: error: satisfaction of constraint 'details::facade_traits<F>::template conv_applicable_ptr<P>' depends on itself
  757 |     details::facade_traits<F>::template conv_applicable_ptr<P> &&
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:757:5: note: while substituting template arguments into constraint expression here
  757 |     details::facade_traits<F>::template conv_applicable_ptr<P> &&
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:819:11: note: while checking the satisfaction of concept 'proxiable<int, Iterator<int>>' requested here
  819 |           proxiable<std::decay_t<P>, F> &&
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:819:11: note: while substituting template arguments into constraint expression here
  819 |           proxiable<std::decay_t<P>, F> &&
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:1995:1: note: while checking constraint satisfaction for template 'proxy<Iterator<int>>' required here
 1995 | ___PRO_BINARY_OP_DISPATCH_IMPL(!=)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:1936:5: note: expanded from macro '___PRO_BINARY_OP_DISPATCH_IMPL'
 1936 |     ___PRO_LHS_OP_DISPATCH_IMPL(BINARY, __VA_ARGS__) \
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:1898:7: note: expanded from macro '___PRO_LHS_OP_DISPATCH_IMPL'
 1898 |       ___PRO_LHS_##TYPE##_OP_DISPATCH_BODY_IMPL(__VA_ARGS__) \
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:60:1: note: expanded from here
   60 | ___PRO_LHS_BINARY_OP_DISPATCH_BODY_IMPL
      | ^
proxy.h:1891:13: note: expanded from macro '___PRO_LHS_BINARY_OP_DISPATCH_BODY_IMPL'
 1891 |             std::forward<T>(self) __VA_ARGS__ std::forward<Arg>(arg))
      |             ^~~
proxy.h:1215:57: note: expanded from macro '___PRO_DIRECT_FUNC_IMPL'
 1215 |     noexcept(noexcept(__VA_ARGS__)) requires(requires { __VA_ARGS__; }) \
      |                                                         ^~~~~~~~~~~
proxy.h:1995:1: note: in instantiation of function template specialization 'pro::proxy<Iterator<int>>::proxy<const int &>' requested here
proxy.h:1936:5: note: expanded from macro '___PRO_BINARY_OP_DISPATCH_IMPL'
 1936 |     ___PRO_LHS_OP_DISPATCH_IMPL(BINARY, __VA_ARGS__) \
      |     ^
proxy.h:1898:7: note: expanded from macro '___PRO_LHS_OP_DISPATCH_IMPL'
 1898 |       ___PRO_LHS_##TYPE##_OP_DISPATCH_BODY_IMPL(__VA_ARGS__) \
      |       ^
<scratch space>:60:1: note: expanded from here
   60 | ___PRO_LHS_BINARY_OP_DISPATCH_BODY_IMPL
      | ^
proxy.h:1891:13: note: expanded from macro '___PRO_LHS_BINARY_OP_DISPATCH_BODY_IMPL'
 1891 |             std::forward<T>(self) __VA_ARGS__ std::forward<Arg>(arg))
      |             ^
proxy.h:1995:1: note: (skipping 39 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
proxy.h:1936:5: note: expanded from macro '___PRO_BINARY_OP_DISPATCH_IMPL'
 1936 |     ___PRO_LHS_OP_DISPATCH_IMPL(BINARY, __VA_ARGS__) \
      |     ^
proxy.h:1898:7: note: expanded from macro '___PRO_LHS_OP_DISPATCH_IMPL'
 1898 |       ___PRO_LHS_##TYPE##_OP_DISPATCH_BODY_IMPL(__VA_ARGS__) \
      |       ^
<scratch space>:60:1: note: expanded from here
   60 | ___PRO_LHS_BINARY_OP_DISPATCH_BODY_IMPL
      | ^
proxy.h:1891:13: note: expanded from macro '___PRO_LHS_BINARY_OP_DISPATCH_BODY_IMPL'
 1891 |             std::forward<T>(self) __VA_ARGS__ std::forward<Arg>(arg))
      |             ^
proxy.h:757:5: note: while substituting template arguments into constraint expression here
  757 |     details::facade_traits<F>::template conv_applicable_ptr<P> &&
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:819:11: note: while checking the satisfaction of concept 'proxiable<int *, Iterator<int>>' requested here
  819 |           proxiable<std::decay_t<P>, F> &&
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:819:11: note: while substituting template arguments into constraint expression here
  819 |           proxiable<std::decay_t<P>, F> &&
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1.cc:21:29: note: while checking constraint satisfaction for template 'proxy<Iterator<int>>' required here
   21 |   pro::proxy<Iterator<int>> p = a;  // <-- Won't compile
      |                             ^
1.cc:21:29: note: in instantiation of function template specialization 'pro::proxy<Iterator<int>>::proxy<int (&)[3]>' requested here
In file included from 1.cc:1:
proxy.h:757:5: error: satisfaction of constraint 'details::facade_traits<F>::template conv_applicable_ptr<P>' depends on itself
  757 |     details::facade_traits<F>::template conv_applicable_ptr<P> &&
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:757:5: note: while substituting template arguments into constraint expression here
  757 |     details::facade_traits<F>::template conv_applicable_ptr<P> &&
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:819:11: note: while checking the satisfaction of concept 'proxiable<int *, Iterator<int>>' requested here
  819 |           proxiable<std::decay_t<P>, F> &&
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:819:11: note: while substituting template arguments into constraint expression here
  819 |           proxiable<std::decay_t<P>, F> &&
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:1995:1: note: while checking constraint satisfaction for template 'proxy<Iterator<int>>' required here
 1995 | ___PRO_BINARY_OP_DISPATCH_IMPL(!=)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:1936:5: note: expanded from macro '___PRO_BINARY_OP_DISPATCH_IMPL'
 1936 |     ___PRO_LHS_OP_DISPATCH_IMPL(BINARY, __VA_ARGS__) \
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:1898:7: note: expanded from macro '___PRO_LHS_OP_DISPATCH_IMPL'
 1898 |       ___PRO_LHS_##TYPE##_OP_DISPATCH_BODY_IMPL(__VA_ARGS__) \
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:60:1: note: expanded from here
   60 | ___PRO_LHS_BINARY_OP_DISPATCH_BODY_IMPL
      | ^
proxy.h:1891:13: note: expanded from macro '___PRO_LHS_BINARY_OP_DISPATCH_BODY_IMPL'
 1891 |             std::forward<T>(self) __VA_ARGS__ std::forward<Arg>(arg))
      |             ^~~
proxy.h:1215:57: note: expanded from macro '___PRO_DIRECT_FUNC_IMPL'
 1215 |     noexcept(noexcept(__VA_ARGS__)) requires(requires { __VA_ARGS__; }) \
      |                                                         ^~~~~~~~~~~
proxy.h:1995:1: note: in instantiation of function template specialization 'pro::proxy<Iterator<int>>::proxy<int *const &>' requested here
proxy.h:1936:5: note: expanded from macro '___PRO_BINARY_OP_DISPATCH_IMPL'
 1936 |     ___PRO_LHS_OP_DISPATCH_IMPL(BINARY, __VA_ARGS__) \
      |     ^
proxy.h:1898:7: note: expanded from macro '___PRO_LHS_OP_DISPATCH_IMPL'
 1898 |       ___PRO_LHS_##TYPE##_OP_DISPATCH_BODY_IMPL(__VA_ARGS__) \
      |       ^
<scratch space>:60:1: note: expanded from here
   60 | ___PRO_LHS_BINARY_OP_DISPATCH_BODY_IMPL
      | ^
proxy.h:1891:13: note: expanded from macro '___PRO_LHS_BINARY_OP_DISPATCH_BODY_IMPL'
 1891 |             std::forward<T>(self) __VA_ARGS__ std::forward<Arg>(arg))
      |             ^
proxy.h:1995:1: note: (skipping 17 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
proxy.h:1936:5: note: expanded from macro '___PRO_BINARY_OP_DISPATCH_IMPL'
 1936 |     ___PRO_LHS_OP_DISPATCH_IMPL(BINARY, __VA_ARGS__) \
      |     ^
proxy.h:1898:7: note: expanded from macro '___PRO_LHS_OP_DISPATCH_IMPL'
 1898 |       ___PRO_LHS_##TYPE##_OP_DISPATCH_BODY_IMPL(__VA_ARGS__) \
      |       ^
<scratch space>:60:1: note: expanded from here
   60 | ___PRO_LHS_BINARY_OP_DISPATCH_BODY_IMPL
      | ^
proxy.h:1891:13: note: expanded from macro '___PRO_LHS_BINARY_OP_DISPATCH_BODY_IMPL'
 1891 |             std::forward<T>(self) __VA_ARGS__ std::forward<Arg>(arg))
      |             ^
proxy.h:757:5: note: while substituting template arguments into constraint expression here
  757 |     details::facade_traits<F>::template conv_applicable_ptr<P> &&
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:819:11: note: while checking the satisfaction of concept 'proxiable<int *, Iterator<int>>' requested here
  819 |           proxiable<std::decay_t<P>, F> &&
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
proxy.h:819:11: note: while substituting template arguments into constraint expression here
  819 |           proxiable<std::decay_t<P>, F> &&
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1.cc:21:29: note: while checking constraint satisfaction for template 'proxy<Iterator<int>>' required here
   21 |   pro::proxy<Iterator<int>> p = a;  // <-- Won't compile
      |                             ^
1.cc:21:29: note: in instantiation of function template specialization 'pro::proxy<Iterator<int>>::proxy<int (&)[3]>' requested here
1.cc:21:29: error: no viable conversion from 'int[3]' to 'pro::proxy<Iterator<int>>'
   21 |   pro::proxy<Iterator<int>> p = a;  // <-- Won't compile
      |                             ^   ~
proxy.h:789:3: note: candidate constructor not viable: no known conversion from 'int[3]' to 'std::nullptr_t' for 1st argument
  789 |   proxy(std::nullptr_t) noexcept : proxy() {}
      |   ^     ~~~~~~~~~~~~~~
proxy.h:801:3: note: candidate constructor not viable: no known conversion from 'int[3]' to 'proxy<Iterator<int>> &&' for 1st argument
  801 |   proxy(proxy&& rhs)
      |   ^     ~~~~~~~~~~~
proxy.h:817:3: note: candidate template ignored: constraints not satisfied [with P = int (&)[3]]
  817 |   proxy(P&& ptr) noexcept(std::is_nothrow_constructible_v<std::decay_t<P>, P>)
      |   ^
proxy.h:819:11: note: because substituted constraint expression is ill-formed: constraint depends on a previously diagnosed expression
  819 |           proxiable<std::decay_t<P>, F> &&
      |           ^
proxy.h:790:3: note: candidate constructor not viable: constraints not satisfied
  790 |   proxy(const proxy&) noexcept requires(F::constraints.copyability ==
      |   ^
proxy.h:790:41: note: because 'Iterator<int>::constraints.copyability == constraint_level::trivial' evaluated to false
  790 |   proxy(const proxy&) noexcept requires(F::constraints.copyability ==
      |                                         ^
proxy.h:792:3: note: candidate constructor not viable: constraints not satisfied
  792 |   proxy(const proxy& rhs)
      |   ^
proxy.h:794:16: note: because 'Iterator<int>::constraints.copyability == constraint_level::nontrivial' evaluated to false
  794 |       requires(F::constraints.copyability == constraint_level::nontrivial ||
      |                ^
proxy.h:795:11: note: and 'Iterator<int>::constraints.copyability == constraint_level::nothrow' evaluated to false
  795 |           F::constraints.copyability == constraint_level::nothrow) {
      |           ^
proxy.h:823:12: note: explicit constructor is not a candidate
  823 |   explicit proxy(std::in_place_type_t<P>, Args&&... args)
      |            ^
3 errors generated.
```

--- a/proxy.h
+++ b/proxy.h
@@ -332,11 +332,19 @@ struct overload_traits<R(Args...) const&& noexcept>
     : overload_traits_impl<qualifier_type::const_rv, true, R, Args...> {};
 
 template <class O>
-struct overload_substitution_traits : inapplicable_traits
-    { template <class> using type = O; };
+struct overload_substitution_traits : inapplicable_traits {
+  template <class> using type = O;
+  template <bool IsDirect, class D, class P>
+  static constexpr bool applicable_ptr =
+      overload_traits<O>::template applicable_ptr<IsDirect, D, P>;
+};
 template <template <class> class O>
 struct overload_substitution_traits<facade_aware_overload_t<O>>
-    : applicable_traits { template <class F> using type = O<F>; };
+    : applicable_traits {
+  template <class F> using type = O<F>;
+  template <bool IsDirect, class D, class P>
+  static constexpr bool applicable_ptr = true;
+};
 template <class O, class F>
 using substituted_overload_t =
     typename overload_substitution_traits<O>::template type<F>;
@@ -393,7 +401,7 @@ struct conv_traits_impl<C, F, Os...> : applicable_traits {
 
   template <class P>
   static constexpr bool applicable_ptr =
-      (overload_traits<substituted_overload_t<Os, F>>::template applicable_ptr<
+      (overload_substitution_traits<Os>::template applicable_ptr<
           C::is_direct, typename C::dispatch_type, P> && ...);
 };
 template <class C, class F> struct conv_traits : inapplicable_traits {};

--- a/tests/proxy_regression_tests.cpp
+++ b/tests/proxy_regression_tests.cpp
@@ -2,7 +2,10 @@
 // Licensed under the MIT License.
 
 #include <gtest/gtest.h>
+#include <vector>
 #include "proxy.h"
+
+namespace proxy_regression_tests_details {
 
 template <class It, class F>
 bool operator==(const It& it, const pro::proxy<F>& rhs) noexcept
@@ -10,17 +13,24 @@ bool operator==(const It& it, const pro::proxy<F>& rhs) noexcept
   return typeid(It) == proxy_typeid(rhs) && it == proxy_cast<const It&>(rhs);
 }
 
-namespace proxy_regression_tests_details {
-
 template <class F>
 using SelfComparisonOverload = bool(const pro::proxy<F>& rhs) const noexcept;
 
 template <class T>
 struct Iterator : pro::facade_builder
     ::support_direct_rtti
+    ::restrict_layout<4 * sizeof(void*)>
     ::add_direct_convention<pro::operator_dispatch<"++">, void() noexcept>
     ::add_direct_convention<pro::operator_dispatch<"!=">, pro::facade_aware_overload_t<SelfComparisonOverload>>
     ::add_convention<pro::implicit_conversion_dispatch, T&() const noexcept>
+    ::build {};
+
+PRO_DEF_MEM_DISPATCH(MemBegin, begin);
+PRO_DEF_MEM_DISPATCH(MemEnd, end);
+template <class T>
+struct Range : pro::facade_builder
+    ::add_convention<MemBegin, pro::proxy<Iterator<T>>()>
+    ::template add_convention<MemEnd, pro::proxy<Iterator<T>>()>
     ::build {};
 
 }  // namespace proxy_regression_tests_details
@@ -43,9 +53,11 @@ TEST(ProxyRegressionTests, TestUnexpectedCompilerWarning) {
 
 // https://github.com/microsoft/proxy/issues/254
 TEST(ProxyRegressionTests, TestProxiableSelfDependency) {
-  int a[3]{1, 2, 3};
-  pro::proxy<details::Iterator<int>> p = a;
-  ASSERT_EQ(*p, 1);
-  ++p;
-  ASSERT_EQ(*p, 2);
+  std::vector<int> original{1, 2, 123};
+  std::vector<int> expected;
+  pro::proxy<details::Range<int>> p = &original;
+  for (int i : *p) {
+    expected.push_back(i);
+  }
+  EXPECT_EQ(expected, original);
 }

--- a/tests/proxy_regression_tests.cpp
+++ b/tests/proxy_regression_tests.cpp
@@ -4,6 +4,29 @@
 #include <gtest/gtest.h>
 #include "proxy.h"
 
+template <class It, class F>
+bool operator==(const It& it, const pro::proxy<F>& rhs) noexcept
+    requires(!std::is_same_v<It, pro::proxy<F>>) {
+  return typeid(It) == proxy_typeid(rhs) && it == proxy_cast<const It&>(rhs);
+}
+
+namespace proxy_regression_tests_details {
+
+template <class F>
+using SelfComparisonOverload = bool(const pro::proxy<F>& rhs) const noexcept;
+
+template <class T>
+struct Iterator : pro::facade_builder
+    ::support_direct_rtti
+    ::add_direct_convention<pro::operator_dispatch<"++">, void() noexcept>
+    ::add_direct_convention<pro::operator_dispatch<"!=">, pro::facade_aware_overload_t<SelfComparisonOverload>>
+    ::add_convention<pro::implicit_conversion_dispatch, T&() const noexcept>
+    ::build {};
+
+}  // namespace proxy_regression_tests_details
+
+namespace details = proxy_regression_tests_details;
+
 // https://github.com/microsoft/proxy/issues/213
 TEST(ProxyRegressionTests, TestUnexpectedCompilerWarning) {
   struct MyTrivialFacade : pro::facade_builder
@@ -16,4 +39,13 @@ TEST(ProxyRegressionTests, TestUnexpectedCompilerWarning) {
   pro::proxy<MyTrivialFacade> p = pro::make_proxy<MyTrivialFacade>([&] { side_effect = 1; });
   (*p)();
   EXPECT_EQ(side_effect, 1);
+}
+
+// https://github.com/microsoft/proxy/issues/254
+TEST(ProxyRegressionTests, TestProxiableSelfDependency) {
+  int a[3]{1, 2, 3};
+  pro::proxy<details::Iterator<int>> p = a;
+  ASSERT_EQ(*p, 1);
+  ++p;
+  ASSERT_EQ(*p, 2);
 }

